cmd/soak-validator/soak-validator.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/amazon-cloudwatch-agent-test/internal/awsservice"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
+	"github.com/shirou/gopsutil/v3/process"
+)
+
+var (
+	cpuLimit = flag.Int("cpuLimit", 1, "agent's upper cpu usage limit in percent")
+	memLimit = flag.Int("memLimit", 200000000, "agent's upper memory usage limit in Bytes")
+	interval = flag.Duration("interval", 10*time.Second, "how frequently to validate")
+	testName = flag.String("testName", "SoakTestLinux", "namespace for metrics")
+)
+
+// main() runs forever.
+// It checks the resource usage of the agent process and reports metrics to CloudWatch.
+// Intentionally not relying on the agent to check itself.
+func main() {
+	flag.Parse()
+	log.Printf("Start validating, cpuLimit %d, memLimit %d MiB", *cpuLimit, *memLimit)
+	// Sleep to avoid alarming on CWA using more resources during start up.
+	time.Sleep(time.Minute)
+	ticker := time.NewTicker(*interval)
+	for range ticker.C {
+		p := getAgentProcess()
+		if p == nil {
+			log.Printf("error: agent process not found")
+			awsservice.ReportMetric(*testName, "FailCount", 1, types.StandardUnitCount)
+			continue
+		}
+		// todo: check for unexpected crash/restart (createtime change).
+		validate("AgentCpu", getCpuUsage(p), float64(*cpuLimit))
+		validate("AgentMemory", getMem(p), float64(*memLimit))
+		// todo: permission denied trying to get numFD, may need to with sudo.
+		//validate("AgentNumFD", getNumFDs(p), 100)
+		validate("AgentNumThreads", getNumThreads(p), 100)
```

Comment:
Did not add an instance-id dimension.
The "namespace" is basically just a dimension, and my intention is to have separate namesapce for each test-type and OS. And I don't imagine there being more than 1 host running the soak test for each of these:
- SoakTestLowLinux
- SoakTestHighLinux
- SoakTestLowWindows
- SoakTestHighWindows
- etc...

If the test is failing, it will be clear from the namespace which host is failing.
At that point the developer can SSH into the instance using SSH key pair or EC2 connect and check the workload, adjust it, etc.